src/runlib.rs

```diff
@@ -1 +1,169 @@
 //! A tool that functionaries can use to create link metadata about a step.
+
+use std::collections::BTreeMap;
+use std::fs::{metadata, File};
+use std::io::{self, BufReader, Write};
+use std::process::Command;
+use walkdir::WalkDir;
+
+use crate::models::{Link, TargetDescription};
+use crate::{Error, Result};
+use crate::{
+    crypto,
+    crypto::PrivateKey,
+    models::{LinkMetadataBuilder, VirtualTargetPath},
+};
+
+/// recordArtifacts is a function that traverses through the passed slice of paths, hashes the content of files
+/// encountered, and returns the path and hashed content in BTreeMap format, wrapped in Result.
+/// If a step in record_artifact fails, the error is returned.
+pub fn record_artifacts(
+    paths: &[&str],
+    // hash_algorithms: Option<&[&str]>,
+) -> Result<BTreeMap<VirtualTargetPath, TargetDescription>> {
+
+    // Initialize artifacts
+    let mut artifacts: BTreeMap<VirtualTargetPath, TargetDescription> = BTreeMap::new();
+
+    // For each path provided, walk the directory and add all files to artifacts
+    for path in paths {
+        for entry in WalkDir::new(path) {
+            let entry = match entry {
+                Ok(content) => content,
+                Err(error) => return Err(Error::from(io::Error::new(
+                    std::io::ErrorKind::Other, format!("Walkdir Error: {}", error)
+                )))
+            };
+            let entry_path = entry.path();
+
+            // TODO: Handle soft/symbolic links, by default is they are ignored, but we should visit them just once
+            // TODO: Handle hidden files & directories
+
+            // If entry is a file, open and hash the file
+            let md = metadata(entry_path)?;
+            if md.is_file() {
+                let file = File::open(entry_path)?;
+                let mut reader = BufReader::new(file);
+                // TODO: handle optional hash_algorithms input
+                let (_length, hashes) =
+                    crypto::calculate_hashes(&mut reader, &[crypto::HashAlgorithm::Sha256])?;
+                let path = entry_path.to_str().unwrap().to_string().replace("./", "");
+                artifacts.insert(VirtualTargetPath::new(path)?, hashes);
+            }
+        }
+    }
+    Ok(artifacts)
+}
+
+/// run_command is a function that, given command arguments, executes commands on a software supply chain step
+/// and returns the stdout and stderr as byproducts.
+/// The first element of cmd_args is used as executable and the rest as command arguments.
+/// If a commands in run_command fails to execute, the error is returned.
+pub fn run_command(
+    cmd_args: &[&str],
+    // TODO run_dir: Option<&str>
```

Comment:
https://doc.rust-lang.org/std/process/struct.Command.html#method.current_dir

I think this may help with this.